Improvements to the tutorial

thomashan opened the pull request:
I felt that there were improvements to be made to the tutorial.
- handling multiple subscribers
- sink that writes back to the kafka
- writing to multiple kafka topics

I have made improvements to cater for these in the pull request.

thomashan committed:
added improvements I wanted to make

thomashan committed:
renamed the reader and the writer

thomashan committed:
renamed the reader and the writer

thomashan committed:
* added broadcasting source
* publish to multiple kafka topics

kciesielski commented:
Very good idea in general, because currently we have quite a few questions regarding broadcasting. Please see my comments for individual code parts.

thomashan committed:
change the readme

thomashan committed:
kafka alert reader coordinator

thomashan committed:
"com.typesafe.akka" %% "akka-stream-experimental" % "2.0-M2" should be pulled in as part of "com.softwaremill.reactivekafka" %% "reactive-kafka-core" % "0.8.1"
We should get rid of the explicit dependency once reactive-kafka-core is updated to use the latest akka-stream

thomashan commented on build.sbt:
```diff
@@ -8,6 +8,7 @@ scalaVersion := "2.11.7"
 
 libraryDependencies ++= Seq(
   "com.softwaremill.reactivekafka" %% "reactive-kafka-core" % "0.8.1",
+  "com.typesafe.akka" %% "akka-stream-experimental" % "2.0-M1",
```
akka-stream-experimental:2.0-M2 should pull in as part of reactive-kafka-core
Can you make changes to reactive-kafka-core so it pulls in the latest akka-stream?

thomashan commented on src/main/scala/sample/reactivekafka/KafkaReaderCoordinator.scala:
```diff
@@ -41,19 +41,53 @@ class KafkaReaderCoordinator(mat: Materializer, topicName: String) extends Actor
+    currencyBroadcaster
       .withAttributes(ActorAttributes.supervisionStrategy(processingDecider))
-      .to(consumerWithOffsetSink.offsetCommitSink).run()
+      .to(consumerWithOffsetSink.offsetCommitSink)
```
I'm writing three streams explicitly to show that these are three explicit streams.
Your way of writing the flow uses the DSL to describe what the flow is.

For me this is a big turn off because
- DSL adds another layer of indirection which I have to learn
- the flow has to be specified beforehand, which is not plausible for non-closed flows
- I wanted an unbounded number of sinks to be able to subscribe to the publisher

From a beginner's point of view I view separating these into three separate streams to be more readable and more comprehensible than the scala DSL you have provided.

thomashan commented on src/main/scala/sample/reactivekafka/KafkaReaderCoordinator.scala:
```diff
@@ -41,19 +41,53 @@ class KafkaReaderCoordinator(mat: Materializer, topicName: String) extends Actor
+    // commit offset so we don't process the kafka message again
+    currencyBroadcaster
       .withAttributes(ActorAttributes.supervisionStrategy(processingDecider))
-      .to(consumerWithOffsetSink.offsetCommitSink).run()
```
Correct me if I'm wrong.
So to guarantee once only delivery you would have to acknowledge that you have received the message.
This is irrespective of whether you have processed it or now.

The intention was to acknowledge the message after broadcast so that any subsequent streams can process the message once.

Is there are better way of doing this?

kciesielski replied:
I understand your flow as:
1. Read from the source (currencyBroadcaster)
2. Write to commit sink 
3. Do more stuff with the msg (in other streams)

If we resigned from manual commit, the result would be actually the same:
1. Read from the source (`currencyBroadcaster`)
2. (Automatically commit) <- this happens automatically under the hood in Kafka Client
3. Do more stuff with the msg

As far as I understand, the "at least once delivery" semantics gives us possibility to make sure that we can put our message through some internal processing and then commit it as "delivered". Without any internal processing between reading and committing, it doesn't give us much.

thomashan replied:
Oh. I didn't realise the kafka client actually did the auto commit.
In which case then, committing the offset will be useless.
So the first flow is effectively going to useless then.

I will take the first flow out then.

kciesielski replied:
Right, auto-commit is the default behavior for Kafka client. We added manual commit in order to allow ensuring that a message gets committed only if passes some internal processes.

thomashan replied:
Ah right.
So the manual commit allow the message to be committed only if the condition is meet.
This would be very useful feature indeed.
May be I need to create another flow to demonstrate this.

thomashan committed:
use options to pass the currency

thomashan commented on src/main/scala/sample/reactivekafka/Coordinator.scala:
```diff
@@ -27,14 +28,18 @@ class Coordinator extends Actor with ActorLogging {
-      reader = Some(context.actorOf(Props(new KafkaReaderCoordinator(materializer, topicName))))
+      currencyWriter = Some(context.actorOf(Props(new KafkaWriterCoordinator(materializer, topicName))))
+      currencyReader = Some(context.actorOf(Props(new KafkaReaderCoordinator(materializer, topicName))))
+      alertReader = Some(context.actorOf(Props(new KafkaAlertReaderCoordinator(materializer, topicName, None))))
```
Options! Yeay!

thomashan committed:
take out the offset commit as it was not giving any value since the offset it committed automatically

thomashan commented on src/main/scala/sample/reactivekafka/KafkaReaderCoordinator.scala:
```diff
@@ -41,19 +41,46 @@ class KafkaReaderCoordinator(mat: Materializer, topicName: String) extends Actor
-    Source(consumerWithOffsetSink.publisher)
-      .map(processMessage)
-      .withAttributes(ActorAttributes.supervisionStrategy(processingDecider))
-      .to(consumerWithOffsetSink.offsetCommitSink).run()
```
So in which scenarios is the offsetCommitSink going to be useful??

kciesielski replied:
Yes, because theoretically 'processMessage' can fail. In this example it will obviously always pass but in real life scenarios something bad may happen. Maybe a better example would be something that randomly fails in the `processMessage` function.

thomashan replied:
I've been thinking about the manual committing issue.
I don't think the manual commit would be useful in this case because it's emitting to multiple sinks.
The previous example was manually committing every message so I don't think it's doing any worse than it was previously doing.

kciesielski commented:
Closing due to #5
